segfault in fast_path/test_callbackq.remove/mt_10/0 on arm

http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3539/label=ARM_test_node,worker=3/console
```
20:30:10 [ RUN      ] fast_path/test_callbackq.remove/mt_10/0
20:30:10 [juno000:93567:0] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x14005d4b0)
20:30:10 
20:30:10 /scrap/jenkins/workspace/gh-ucx-pr/label/ARM_test_node/worker/3/contrib/../src/ucs/datastruct/callbackq.c: [ ucs_callbackq_remove_common() ]
20:30:10       ...
20:30:10       175 
20:30:10       176     /* replace removed with last */
20:30:10       177     elems[idx] = elems[last_idx];
20:30:10 ==>   178     ucs_callbackq_elem_reset(cbq, &elems[last_idx]);
20:30:10       179 
20:30:10       180     if (*remove_mask & UCS_BIT(last_idx)) {
20:30:10       181         /* replaced by marked-for-removal element, still need to remove 'idx' */
20:30:10 
20:30:10 
==== backtrace ====
20:30:10  0 0x000000000003e678 ucs_callbackq_remove_common()  /scrap/jenkins/workspace/gh-ucx-pr/label/ARM_test_node/worker/3/contrib/../src/ucs/datastruct/callbackq.c:178
20:30:10  1 0x000000000003e678 ucs_callbackq_remove_fast()  /scrap/jenkins/workspace/gh-ucx-pr/label/ARM_test_node/worker/3/contrib/../src/ucs/datastruct/callbackq.c:225
20:30:10  2 0x000000000003e678 ucs_callbackq_purge_fast()  /scrap/jenkins/workspace/gh-ucx-pr/label/ARM_test_node/worker/3/contrib/../src/ucs/datastruct/callbackq.c:240
20:30:10  3 0x000000000003ec30 ucs_callbackq_slow_proxy()  /scrap/jenkins/workspace/gh-ucx-pr/label/ARM_test_node/worker/3/contrib/../src/ucs/datastruct/callbackq.c:392
20:30:10  4 0x000000000068690c test_callbackq_remove_Test::test_body()  /scrap/jenkins/workspace/gh-ucx-pr/label/ARM_test_node/worker/3/contrib/../src/ucs/datastruct/callbackq.h:168
20:30:10  5 0x000000000068690c test_callbackq::dispatch()  /scrap/jenkins/workspace/gh-ucx-pr/label/ARM_test_node/worker/3/contrib/../test/gtest/ucs/test_callbackq.cc:113
20:30:10  6 0x000000000068690c test_callbackq_remove_Test::test_body()  /scrap/jenkins/workspace/gh-ucx-pr/label/ARM_test_node/worker/3/contrib/../test/gtest/ucs/test_callbackq.cc:276
20:30:10  7 0x000000000047a80c ucs::test_base::thread_func()  /scrap/jenkins/workspace/gh-ucx-pr/label/ARM_test_node/worker/3/contrib/../test/gtest/common/test.cc:268
20:30:10  8 0x0000000000007bb0 start_thread()  :0
20:30:10  9 0x00000000000da680 thread_start()  :0
20:30:10 ===================
...
21:29:03 Build timed out (after 100 minutes). Marking the build as failed.
21:29:04 make: *** [test] Terminated
```

--- a/src/ucs/datastruct/callbackq.h
+++ b/src/ucs/datastruct/callbackq.h
@@ -152,6 +152,7 @@ void ucs_callbackq_remove_safe(ucs_callbackq_t *cbq, int id);
 
 /**
  * Dispatch callbacks from the callback queue.
+ * Must be called from single thread only.
  *
  * @param  [in] cbq      Callback queue to dispatch callbacks from.
 

--- a/test/gtest/ucs/test_callbackq.cc
+++ b/test/gtest/ucs/test_callbackq.cc
@@ -263,31 +263,40 @@ UCS_MT_TEST_P(test_callbackq, remove, 10) {
 
     init_ctx(&ctx1);
 
-    if (barrier()) {
+    if (barrier()) /*1*/ {
         add_safe(&ctx1);
-    }
-
-    sched_yield();
-
-    if (barrier()) {
-        callback_ctx ctx2;
-        init_ctx(&ctx2);
-
-        dispatch(1000);
-        add(&ctx2);
-        barrier();
+        dispatch(100);
+        barrier(); /*2*/
         remove_safe(&ctx1);
-        remove_safe(&ctx2);
         dispatch(1);
 
-        /* this should remove all instances of 'ctx', including queued async
-         * commands */
         uint32_t count1 = ctx1.count;
         dispatch(100);
         EXPECT_EQ(count1, ctx1.count);
-    } else {
+
+        barrier();/*3*/
+        dispatch(1); /* will remove ctx2 on other threads */
+        barrier();/*4*/
+
+        barrier();/*5*/
         dispatch(100);
-        barrier();
+        barrier();/*6*/
+    } else {
+        callback_ctx ctx2;
+        init_ctx(&ctx2);
+        add_safe(&ctx2);
+        barrier(); /*2*/
+
+        /* ask to ctx2 and wait until the main thread actually removes it */
+        remove_safe(&ctx2);
+        barrier(); /*3*/
+        barrier(); /*4*/
+
+        /* make sure ctx2 is not dispatched */
+        uint32_t count1 = ctx2.count;
+        barrier();/*5*/
+        barrier();/*6*/
+        EXPECT_EQ(count1, ctx2.count);
     }
 }
 

